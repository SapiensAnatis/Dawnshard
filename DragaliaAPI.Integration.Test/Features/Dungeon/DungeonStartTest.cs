using DragaliaAPI.Models.Generated;
using DragaliaAPI.Shared.Definitions.Enums;
using Microsoft.EntityFrameworkCore;
using Snapshooter;
using Snapshooter.Xunit;

namespace DragaliaAPI.Integration.Test.Features.Dungeon;

/// <summary>
/// Tests <see cref="DragaliaAPI.Features.Dungeon.Start.DungeonStartController"/>.
/// </summary>
public class DungeonStartTest : TestFixture
{
    public DungeonStartTest(CustomWebApplicationFactory factory, ITestOutputHelper outputHelper)
        : base(factory, outputHelper)
    {
        ImportSave();
    }

    [Fact]
    public async Task Start_OneTeam_HasExpectedPartyUnitList()
    {
        DungeonStartStartData response = (
            await Client.PostMsgpack<DungeonStartStartData>(
                "/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    party_no_list = new List<int>() { 1 },
                    quest_id = 100010103
                }
            )
        ).data;

        Snapshot.Match(response.ingame_data.party_info.party_unit_list, SnapshotOptions);

        response.ingame_data.party_info.party_unit_list.Should().HaveCount(4);
        response.ingame_data.party_info.party_unit_list
            .Should()
            .BeInAscendingOrder(x => x.position);
        response.ingame_data.party_info.party_unit_list
            .Should()
            .OnlyHaveUniqueItems(x => x.position);
    }

    [Fact]
    public async Task Start_TwoTeams_HasExpectedPartyUnitList()
    {
        DungeonStartStartData response = (
            await Client.PostMsgpack<DungeonStartStartData>(
                "/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    party_no_list = new List<int>() { 37, 38 },
                    quest_id = 100010103
                }
            )
        ).data;

        // Abuse of snapshots here is lazy, but the resulting JSON is thousands of lines long...
        Snapshot.Match(response.ingame_data.party_info.party_unit_list, SnapshotOptions);

        response.ingame_data.party_info.party_unit_list.Should().HaveCount(8);
        response.ingame_data.party_info.party_unit_list
            .Should()
            .BeInAscendingOrder(x => x.position);
        response.ingame_data.party_info.party_unit_list
            .Should()
            .OnlyHaveUniqueItems(x => x.position);
    }

    [Fact]
    public async Task Start_WeaponPassivesUnlocked_IncludedInPartyUnitList()
    {
        DungeonStartStartData response = (
            await Client.PostMsgpack<DungeonStartStartData>(
                "/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    party_no_list = new List<int>() { 38 },
                    quest_id = 100010103
                }
            )
        ).data;

        response.ingame_data.party_info.party_unit_list
            .First(x => x.chara_data!.chara_id == Charas.GalaMascula)
            .game_weapon_passive_ability_list.Should()
            .Contain(x => x.weapon_passive_ability_id == 1020211);
    }

    [Fact]
    public async Task StartAssignUnit_HasExpectedPartyList()
    {
        DungeonSkipStartAssignUnitRequest request =
            new()
            {
                quest_id = 100010103,
                request_party_setting_list = new List<PartySettingList>()
                {
                    new()
                    {
                        unit_no = 1,
                        chara_id = Charas.GalaLeonidas,
                        equip_weapon_body_id = WeaponBodies.Draupnir,
                        equip_dragon_key_id = (ulong)GetDragonKeyId(Dragons.Horus),
                        equip_crest_slot_type_1_crest_id_1 = AbilityCrests.PrimalCrisis,
                        equip_crest_slot_type_1_crest_id_2 = AbilityCrests.TheCutieCompetition,
                        equip_crest_slot_type_1_crest_id_3 = AbilityCrests.AnIndelibleDate,
                        equip_crest_slot_type_2_crest_id_1 = AbilityCrests.BeautifulGunman,
                        equip_crest_slot_type_2_crest_id_2 = AbilityCrests.DragonArcanum,
                        equip_talisman_key_id = (ulong)GetTalismanKeyId(Talismans.GalaLeonidas),
                        equip_crest_slot_type_3_crest_id_1 = AbilityCrests.AKnightsDreamAxesBoon,
                        equip_crest_slot_type_3_crest_id_2 = AbilityCrests.CrownofLightSerpentsBoon,
                        edit_skill_1_chara_id = Charas.GalaZethia,
                        edit_skill_2_chara_id = Charas.GalaMascula,
                    },
                    new()
                    {
                        unit_no = 2,
                        chara_id = Charas.GalaGatov,
                        equip_weapon_body_id = WeaponBodies.Mjoelnir,
                        equip_dragon_key_id = (ulong)GetDragonKeyId(Dragons.GalaMars),
                        equip_crest_slot_type_1_crest_id_1 = AbilityCrests.TheCutieCompetition,
                        equip_crest_slot_type_1_crest_id_2 = AbilityCrests.KungFuMasters,
                        equip_crest_slot_type_1_crest_id_3 = AbilityCrests.BondsBetweenWorlds,
                        equip_crest_slot_type_2_crest_id_1 = AbilityCrests.DragonArcanum,
                        equip_crest_slot_type_2_crest_id_2 = AbilityCrests.BeautifulNothingness,
                        equip_talisman_key_id = (ulong)GetTalismanKeyId(Talismans.GalaMym),
                        equip_crest_slot_type_3_crest_id_1 =
                            AbilityCrests.TutelarysDestinyWolfsBoon,
                        equip_crest_slot_type_3_crest_id_2 =
                            AbilityCrests.TestamentofEternityFishsBoon,
                    }
                }
            };

        DungeonStartStartAssignUnitData response = (
            await Client.PostMsgpack<DungeonStartStartAssignUnitData>(
                "/dungeon_start/start_assign_unit",
                request
            )
        ).data;

        // Only test the first two since the others are empty
        Snapshot.Match(response.ingame_data.party_info.party_unit_list.Take(2), SnapshotOptions);

        response.ingame_data.party_info.party_unit_list.Should().HaveCount(4);
        response.ingame_data.party_info.party_unit_list
            .Should()
            .Contain(x => x.chara_data!.chara_id == Charas.GalaLeonidas)
            .And.Contain(x => x.chara_data!.chara_id == Charas.GalaGatov);
    }

    private static readonly Func<MatchOptions, MatchOptions> SnapshotOptions = opts =>
        opts.IgnoreField<long>("$..dragon_data.dragon_key_id")
            .IgnoreField<long>("$..talisman_data.talisman_key_id");
}
