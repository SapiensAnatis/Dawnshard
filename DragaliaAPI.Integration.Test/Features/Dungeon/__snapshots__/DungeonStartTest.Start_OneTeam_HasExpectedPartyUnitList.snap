[
  {
    "position": 1,
    "chara_data": {
      "chara_id": "GalaZena",
      "exp": 8866950,
      "level": 100,
      "additional_max_level": 20,
      "hp": 923,
      "attack": 586,
      "ex_ability_level": 5,
      "ex_ability_2_level": 5,
      "ability_1_level": 3,
      "ability_2_level": 3,
      "ability_3_level": 3,
      "is_new": 1,
      "skill_1_level": 4,
      "skill_2_level": 3,
      "burst_attack_level": 2,
      "rarity": 5,
      "limit_break_count": 5,
      "hp_plus_count": 100,
      "attack_plus_count": 100,
      "status_plus_count": 0,
      "combo_buildup_count": 1,
      "is_unlock_edit_skill": true,
      "gettime": "2020-09-30T06:09:26+00:00",
      "mana_circle_piece_id_list": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50,
        51,
        52,
        53,
        54,
        55,
        56,
        57,
        58,
        59,
        60,
        61,
        62,
        63,
        64,
        65,
        66,
        67,
        68,
        69,
        70
      ],
      "is_temporary": 0,
      "list_view_flag": 1
    },
    "dragon_data": {
      "dragon_id": "JeannedArc",
      "dragon_key_id": 334,
      "level": 120,
      "exp": 3365620,
      "is_lock": true,
      "is_new": true,
      "get_time": "2019-10-13T08:40:10+00:00",
      "skill_1_level": 2,
      "ability_1_level": 6,
      "ability_2_level": 6,
      "limit_break_count": 5,
      "hp_plus_count": 50,
      "attack_plus_count": 50,
      "status_plus_count": 0
    },
    "weapon_skin_data": {
      "weapon_skin_id": 0
    },
    "weapon_body_data": {
      "weapon_body_id": "DelFrith",
      "buildup_count": 80,
      "limit_break_count": 8,
      "limit_over_count": 1,
      "skill_no": 1,
      "skill_level": 2,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "equipable_count": 1,
      "additional_crest_slot_type_1_count": 1,
      "additional_crest_slot_type_2_count": 0,
      "additional_crest_slot_type_3_count": 2,
      "additional_effect_count": 0
    },
    "crest_slot_type_1_crest_list": [
      {
        "ability_crest_id": "ExtremeTeamwork",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "PrimalCrisis",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "SpiritoftheSeason",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_2_crest_list": [
      {
        "ability_crest_id": "FromWhenceHeComes",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 2,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "TheWarrioresses",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 1,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_3_crest_list": [
      {
        "ability_crest_id": "CrownofLightSerpentsBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 3,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      },
      {
        "ability_crest_id": "AKnightsDreamAxesBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      }
    ],
    "talisman_data": {
      "talisman_key_id": 490,
      "talisman_id": "GalaLuca",
      "talisman_ability_id_1": 340000069,
      "talisman_ability_id_2": 340000108,
      "talisman_ability_id_3": 0,
      "additional_hp": 0,
      "additional_attack": 0,
      "is_new": true,
      "is_lock": true,
      "gettime": "2021-12-30T10:10:51+00:00"
    },
    "edit_skill_1_chara_data": {
      "chara_id": "ShaWujing",
      "edit_skill_level": 2
    },
    "edit_skill_2_chara_data": {
      "chara_id": "GalaMym",
      "edit_skill_level": 4
    },
    "game_weapon_passive_ability_list": [],
    "dragon_reliability_level": 30
  },
  {
    "position": 2,
    "chara_data": {
      "chara_id": "Mitsuhide",
      "exp": 1191950,
      "level": 80,
      "additional_max_level": 0,
      "hp": 763,
      "attack": 497,
      "ex_ability_level": 5,
      "ex_ability_2_level": 5,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "ability_3_level": 2,
      "is_new": 1,
      "skill_1_level": 3,
      "skill_2_level": 2,
      "burst_attack_level": 2,
      "rarity": 5,
      "limit_break_count": 4,
      "hp_plus_count": 100,
      "attack_plus_count": 100,
      "status_plus_count": 0,
      "combo_buildup_count": 0,
      "is_unlock_edit_skill": true,
      "gettime": "2020-01-13T08:00:24+00:00",
      "mana_circle_piece_id_list": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50
      ],
      "is_temporary": 0,
      "list_view_flag": 1
    },
    "dragon_data": {
      "dragon_id": "GalaRebornJeanne",
      "dragon_key_id": 400,
      "level": 100,
      "exp": 1240020,
      "is_lock": true,
      "is_new": true,
      "get_time": "2021-09-29T07:23:05+00:00",
      "skill_1_level": 2,
      "ability_1_level": 5,
      "ability_2_level": 5,
      "limit_break_count": 4,
      "hp_plus_count": 50,
      "attack_plus_count": 50,
      "status_plus_count": 0
    },
    "weapon_skin_data": {
      "weapon_skin_id": 0
    },
    "weapon_body_data": {
      "weapon_body_id": "ClaíomhSolais",
      "buildup_count": 80,
      "limit_break_count": 8,
      "limit_over_count": 1,
      "skill_no": 1,
      "skill_level": 2,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "equipable_count": 1,
      "additional_crest_slot_type_1_count": 1,
      "additional_crest_slot_type_2_count": 0,
      "additional_crest_slot_type_3_count": 2,
      "additional_effect_count": 0
    },
    "crest_slot_type_1_crest_list": [
      {
        "ability_crest_id": "ExtremeTeamwork",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "BlossomsontheWatersEdge",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 1,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "SpiritoftheSeason",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_2_crest_list": [
      {
        "ability_crest_id": "SnipersAllure",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "APassionforProduce",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 2,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_3_crest_list": [
      {
        "ability_crest_id": "AKnightsDreamAxesBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      },
      {
        "ability_crest_id": "CrownofLightSerpentsBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 3,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      }
    ],
    "talisman_data": {
      "talisman_key_id": 590,
      "talisman_id": "HalloweenLaxi",
      "talisman_ability_id_1": 340000009,
      "talisman_ability_id_2": 340000076,
      "talisman_ability_id_3": 0,
      "additional_hp": 0,
      "additional_attack": 0,
      "is_new": true,
      "is_lock": true,
      "gettime": "2022-01-18T18:57:35+00:00"
    },
    "edit_skill_1_chara_data": {
      "chara_id": "Cleo",
      "edit_skill_level": 4
    },
    "edit_skill_2_chara_data": {
      "chara_id": "Ranzal",
      "edit_skill_level": 4
    },
    "game_weapon_passive_ability_list": [
      {
        "weapon_passive_ability_id": 1030405
      }
    ],
    "dragon_reliability_level": 30
  },
  {
    "position": 3,
    "chara_data": {
      "chara_id": "BondforgedZethia",
      "exp": 1191950,
      "level": 80,
      "additional_max_level": 0,
      "hp": 751,
      "attack": 510,
      "ex_ability_level": 5,
      "ex_ability_2_level": 5,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "ability_3_level": 2,
      "is_new": 1,
      "skill_1_level": 3,
      "skill_2_level": 2,
      "burst_attack_level": 2,
      "rarity": 5,
      "limit_break_count": 4,
      "hp_plus_count": 100,
      "attack_plus_count": 100,
      "status_plus_count": 0,
      "combo_buildup_count": 0,
      "is_unlock_edit_skill": false,
      "gettime": "2022-03-31T13:48:38+00:00",
      "mana_circle_piece_id_list": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50
      ],
      "is_temporary": 0,
      "list_view_flag": 1
    },
    "dragon_data": {
      "dragon_id": "GalaThor",
      "dragon_key_id": 390,
      "level": 100,
      "exp": 1240020,
      "is_lock": true,
      "is_new": true,
      "get_time": "2020-12-28T06:38:44+00:00",
      "skill_1_level": 2,
      "ability_1_level": 5,
      "ability_2_level": 5,
      "limit_break_count": 4,
      "hp_plus_count": 50,
      "attack_plus_count": 50,
      "status_plus_count": 0
    },
    "weapon_skin_data": {
      "weapon_skin_id": 0
    },
    "weapon_body_data": {
      "weapon_body_id": "Brionac",
      "buildup_count": 80,
      "limit_break_count": 8,
      "limit_over_count": 1,
      "skill_no": 1,
      "skill_level": 2,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "equipable_count": 1,
      "additional_crest_slot_type_1_count": 1,
      "additional_crest_slot_type_2_count": 0,
      "additional_crest_slot_type_3_count": 2,
      "additional_effect_count": 0
    },
    "crest_slot_type_1_crest_list": [
      {
        "ability_crest_id": "ExtremeTeamwork",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "SpiritoftheSeason",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "BloodyLittleQueen",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 1,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_2_crest_list": [
      {
        "ability_crest_id": "ChariotDrift",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "SnipersAllure",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_3_crest_list": [
      {
        "ability_crest_id": "AKnightsDreamAxesBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      },
      {
        "ability_crest_id": "TutelarysDestinyWolfsBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      }
    ],
    "talisman_data": {
      "talisman_key_id": 530,
      "talisman_id": "Pecorine",
      "talisman_ability_id_1": 340000008,
      "talisman_ability_id_2": 340000058,
      "talisman_ability_id_3": 0,
      "additional_hp": 0,
      "additional_attack": 0,
      "is_new": true,
      "is_lock": true,
      "gettime": "2022-01-08T20:49:41+00:00"
    },
    "edit_skill_1_chara_data": {
      "chara_id": "HalloweenLowen",
      "edit_skill_level": 3
    },
    "edit_skill_2_chara_data": {
      "chara_id": "Vixel",
      "edit_skill_level": 4
    },
    "game_weapon_passive_ability_list": [],
    "dragon_reliability_level": 30
  },
  {
    "position": 4,
    "chara_data": {
      "chara_id": "GalaLuca",
      "exp": 3871830,
      "level": 90,
      "additional_max_level": 15,
      "hp": 825,
      "attack": 573,
      "ex_ability_level": 5,
      "ex_ability_2_level": 5,
      "ability_1_level": 3,
      "ability_2_level": 3,
      "ability_3_level": 3,
      "is_new": 1,
      "skill_1_level": 3,
      "skill_2_level": 3,
      "burst_attack_level": 2,
      "rarity": 5,
      "limit_break_count": 5,
      "hp_plus_count": 100,
      "attack_plus_count": 100,
      "status_plus_count": 0,
      "combo_buildup_count": 0,
      "is_unlock_edit_skill": true,
      "gettime": "2020-01-24T20:32:42+00:00",
      "mana_circle_piece_id_list": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50,
        51,
        52,
        53,
        54,
        55,
        56,
        57,
        58,
        59,
        60,
        61,
        62,
        63
      ],
      "is_temporary": 0,
      "list_view_flag": 1
    },
    "dragon_data": {
      "dragon_id": "GalaChronosNyx",
      "dragon_key_id": 408,
      "level": 100,
      "exp": 1240020,
      "is_lock": false,
      "is_new": true,
      "get_time": "2021-10-30T07:53:26+00:00",
      "skill_1_level": 2,
      "ability_1_level": 5,
      "ability_2_level": 5,
      "limit_break_count": 4,
      "hp_plus_count": 50,
      "attack_plus_count": 50,
      "status_plus_count": 0
    },
    "weapon_skin_data": {
      "weapon_skin_id": 0
    },
    "weapon_body_data": {
      "weapon_body_id": "Fragarach",
      "buildup_count": 80,
      "limit_break_count": 8,
      "limit_over_count": 1,
      "skill_no": 1,
      "skill_level": 2,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "equipable_count": 1,
      "additional_crest_slot_type_1_count": 1,
      "additional_crest_slot_type_2_count": 0,
      "additional_crest_slot_type_3_count": 2,
      "additional_effect_count": 0
    },
    "crest_slot_type_1_crest_list": [
      {
        "ability_crest_id": "PecorinesGrandAdventure",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 1,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "PrayersUntoHim",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 1,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "SpiritoftheSeason",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_2_crest_list": [
      {
        "ability_crest_id": "ChariotDrift",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "TheNoblesDayOff",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_3_crest_list": [
      {
        "ability_crest_id": "CrownofLightSerpentsBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 3,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      },
      {
        "ability_crest_id": "AKnightsDreamAxesBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      }
    ],
    "talisman_data": {
      "talisman_key_id": 529,
      "talisman_id": "Pecorine",
      "talisman_ability_id_1": 340000029,
      "talisman_ability_id_2": 340000056,
      "talisman_ability_id_3": 0,
      "additional_hp": 0,
      "additional_attack": 0,
      "is_new": true,
      "is_lock": true,
      "gettime": "2022-01-08T20:49:41+00:00"
    },
    "edit_skill_1_chara_data": {
      "chara_id": "Empty",
      "edit_skill_level": 0
    },
    "edit_skill_2_chara_data": {
      "chara_id": "Vixel",
      "edit_skill_level": 4
    },
    "game_weapon_passive_ability_list": [],
    "dragon_reliability_level": 30
  }
]
